From: ChepKun <ProgrammerKun@gmail.com>
Subject: [PATCH] frameworks/opt: remove sim added dialog

diff --git a/src/java/com/android/internal/telephony/UiccCard.java b/src/java/com/android/internal/telephony/UiccCard.java
index 038a138..8d8023c 100644
--- a/src/java/com/android/internal/telephony/UiccCard.java
+++ b/src/java/com/android/internal/telephony/UiccCard.java
@@ -229,48 +229,7 @@ public class UiccCard {
     }
 
     private void onIccSwap(boolean isAdded) {
-        synchronized (mLock) {
-            // TODO: Here we assume the device can't handle SIM hot-swap
-            //      and has to reboot. We may want to add a property,
-            //      e.g. REBOOT_ON_SIM_SWAP, to indicate if modem support
-            //      hot-swap.
-            DialogInterface.OnClickListener listener = null;
-
-
-            // TODO: SimRecords is not reset while SIM ABSENT (only reset while
-            //       Radio_off_or_not_available). Have to reset in both both
-            //       added or removed situation.
-            listener = new DialogInterface.OnClickListener() {
-                @Override
-                public void onClick(DialogInterface dialog, int which) {
-                    synchronized (mLock) {
-                        if (which == DialogInterface.BUTTON_POSITIVE) {
-                            if (DBG) log("Reboot due to SIM swap");
-                            PowerManager pm = (PowerManager) mContext
-                                    .getSystemService(Context.POWER_SERVICE);
-                            pm.reboot("SIM is added.");
-                        }
-                    }
-                }
-
-            };
-
-            Resources r = Resources.getSystem();
-
-            String title = (isAdded) ? r.getString(R.string.sim_added_title) :
-                r.getString(R.string.sim_removed_title);
-            String message = (isAdded) ? r.getString(R.string.sim_added_message) :
-                r.getString(R.string.sim_removed_message);
-            String buttonTxt = r.getString(R.string.sim_restart_button);
-
-            AlertDialog dialog = new AlertDialog.Builder(mContext)
-            .setTitle(title)
-            .setMessage(message)
-            .setPositiveButton(buttonTxt, listener)
-            .create();
-            dialog.getWindow().setType(WindowManager.LayoutParams.TYPE_SYSTEM_ALERT);
-            dialog.show();
-        }
+	// fix that
     }
 
     protected Handler mHandler = new Handler() {
